CMakeLists.txt

```diff
@@ -36,10 +51,23 @@ if(BUILD_TESTING)
   ament_add_gtest(test_split test/test_split.cpp)
 
   ament_add_gtest(test_filesystem_helper test/test_filesystem_helper.cpp)
+
+  add_library(test_library SHARED test/test_library.cpp)
+  ament_add_gtest(test_find_library test/test_find_library.cpp)
+  target_link_libraries(test_find_library ${PROJECT_NAME} test_library)
+  set_tests_properties(test_find_library PROPERTIES
+    ENVIRONMENT
+      "_TEST_LIBRARY_DIR=$<TARGET_FILE_DIR:test_library>;_TEST_LIBRARY=$<TARGET_FILE:test_library>")
```

Comment:
Sigh... I'm thinking that Windows CMake or something else is maybe failing here?
I've recently gotten a Windows machine (for another purpose), but will see if I can dig in a bit more.

Reply:
That being said, thank you for your patience thus far!

Reply:
Trying to get this setup on Windows, but not being a Windows developer, to try and reproduce this bug is very time consuming and frustrating.

Is there some free Azure container / VM service that has ROS 2 provisioned, that I could possible connect to so that I can debug from there? (For me and other developers, who do not wish to spend time / money setting up Windows environments?)

And can you point me to a unittest that does this sort of thing successfully?
(Something more explicit than saying "Hey, it loaded the library, cool.")

If there isn't one already, I may just shortcut this for now and special-case the Windows logic to pass on CI by acknowledging the current bug (which I'm still confident is CMake, pending any other evidence to the contrary).

Reply:
Currently the test fails with:

```
C:\J\workspace\ci_windows\ws\src\ros2\rcpputils\test\test_find_library.cpp(58): error: Expected equality of these values:
  test_lib_actual
    Which is: "C:/J/workspace/ci_windows/ws/build/rcpputils/Release/test_library.dll"
  test_lib_expected
    Which is: "C:/J/workspace/ci_windows/ws/build/rcpputils/Release"
```

Your test code retrieves two environment variables: first the library path then the library dir. But you are missing the fact that the second call overwrites the first variable. See the API documentation of that function: https://github.com/ros2/rcutils/blob/9af59e871e0bfaadb0cbb32014a7167efb9760c0/include/rcutils/get_env.h#L27-L29

So the value of the first environment variable must be copied to its own memory before fetching the second. See https://github.com/ros2/rcpputils/pull/25/commits/ed7b9fa1a1f757387250aa25b1156383695fb226 for the fix.